mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -0,0 +1,425 @@
+    pub taker_payment_hex: Vec<u8>,
+    pub taker_payment_lock: u64,
+    pub taker_pub: Vec<u8>,
+    pub taker_coin_start_block: u64,
+    pub taker_payment_confirmations: u64,
+    pub taker_payment_requires_nota: Option<bool>,
+    pub taker_amount: BigDecimal,
+    pub maker_coin: String,
+    pub maker_payment_hex: Vec<u8>,
+    pub maker_payment_lock: u32,
+    pub maker_pub: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub struct TakerPaymentSpentData {
+    pub transaction: TransactionIdentifier,
+    pub secret: H256Json,
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
+pub struct WatcherError {
+    error: String,
+}
+
+impl From<String> for WatcherError {
+    fn from(error: String) -> Self { WatcherError { error } }
+}
+
+impl From<&str> for WatcherError {
+    fn from(e: &str) -> Self { WatcherError { error: e.to_owned() } }
+}
+
+#[allow(clippy::large_enum_variant)]
+pub enum RunWatcherInput {
+    StartNew(Watcher),
+}
+
+impl RunWatcherInput {
+    fn uuid(&self) -> &Uuid {
+        match self {
+            RunWatcherInput::StartNew(swap) => &swap.uuid,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+#[serde(tag = "type", content = "data")]
+#[allow(clippy::large_enum_variant)]
+pub enum WatcherEvent {
+    Started,
+    StartFailed(WatcherError),
+    TakerPaymentWaitConfirmFailed(WatcherError),
+    TakerPaymentValidatedAndConfirmed,
+    TakerPaymentValidateFailed(WatcherError),
+    TakerPaymentSpent(TakerPaymentSpentData),
+    TakerPaymentWaitForSpendFailed(WatcherError),
+    MakerPaymentSpendFailed(WatcherError),
+    MakerPaymentSpent(TransactionIdentifier),
+    Finished,
+}
+
+impl WatcherEvent {
+    pub fn status_str(&self) -> String {
+        match self {
+            WatcherEvent::Started => "Started...".to_owned(),
+            WatcherEvent::StartFailed(_) => "Start failed...".to_owned(),
+            WatcherEvent::TakerPaymentWaitConfirmFailed(_) => {
+                "Taker payment wait for confirmation failed...".to_owned()
+            },
+            WatcherEvent::TakerPaymentValidatedAndConfirmed => "Taker payment validated and confirmed...".to_owned(),
+            WatcherEvent::TakerPaymentValidateFailed(_) => "Taker payment validate failed...".to_owned(),
+            WatcherEvent::TakerPaymentSpent(_) => "Taker payment spent...".to_owned(),
+            WatcherEvent::TakerPaymentWaitForSpendFailed(_) => "Taker payment wait for spend failed...".to_owned(),
+            WatcherEvent::MakerPaymentSpendFailed(_) => "Maker payment spend failed...".to_owned(),
+            WatcherEvent::MakerPaymentSpent(_) => "Maker payment spent...".to_owned(),
+            WatcherEvent::Finished => "Finished".to_owned(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum WatcherCommand {
+    Start,
+    ValidateTakerPayment,
+    WaitForTakerPaymentSpend,
+    SpendMakerPayment,
+    Finish,
+}
+
+impl Watcher {
```

Comment:
It should not follow the same design that was used for swap implementation. I even consider this `WatcherMut` approach not good in general - these `r()/w()` calls bloat the code and cause lifetime issues in some cases, forcing us to clone or reassign values bloating the code even more like here:
https://github.com/KomodoPlatform/atomicDEX-API/blob/ebb0e81e14e4bd1e90a8068552c5e072bd5f1791/mm2src/mm2_main/src/lp_swap/maker_swap.rs#L1136

Please consider refactoring this using our state machine pattern https://github.com/KomodoPlatform/atomicDEX-API/blob/ebb0e81e14e4bd1e90a8068552c5e072bd5f1791/mm2src/common/patterns/state_machine.rs

Implementation examples:
https://github.com/KomodoPlatform/atomicDEX-API/blob/ebb0e81e14e4bd1e90a8068552c5e072bd5f1791/mm2src/mm2_net/src/wasm_ws.rs
https://github.com/KomodoPlatform/atomicDEX-API/blob/ebb0e81e14e4bd1e90a8068552c5e072bd5f1791/mm2src/coins/utxo/bch_and_slp_tx_history.rs

Reply:
PS, you can do this in the next iteration/PR 🙂

Reply:
Yeah this will take some time obviously :sweat_smile: